### ISSUE TITLE
Can Dex actually allow question marks in array sizes?

### ISSUE THREAD

**Issue by axch** [google-research/dex-lang#947]
I tried to define a function specialized to a flat array with an anonymous (unknown) size, like so:
```
def fin_mean (xs: (Fin _)=>Float) : Float =
  mean xs
```
and I got this:
```
Not implemented:
Please report this at github.com/google-research/dex-lang/issues

Type inference of this program requires delayed interface resolution

def fin_mean (xs: (Fin _)=>Float) : Float =
                   ^^^^^^^^^^^^^
```

This doesn't block anything for me, because it's easy enough to just add the implicit `Nat` argument to serve as a name for the size, but the gap seems unaesthetic.

**Comment by axch** [comment]
Hm.  Then "not implemented" is probably not a good error message.

**Comment by apaszke** [comment]
I agree it's not good. But putting an extra inference variable in there does hit a not implemented case in inference, so it's also true. We used to support this before safer names, but not anymore.
